sqlmesh/core/plan/definition.py

```diff
@@ -398,28 +432,61 @@ def _missing_intervals(self) -> t.Dict[str, Intervals]:
                     start=self._start or scheduler.earliest_start_date(self.snapshots),
                     end=self._end or now(),
                     latest=self._latest,
+                    is_dev=self.is_dev,
                     restatements=self.restatements,
                 ).items()
             }
 
         return self.__missing_intervals
 
+    @classmethod
+    def _get_end_date(cls, end_and_units: t.List[t.Tuple[int, IntervalUnit]]) -> TimeLike:
+        if end_and_units:
+            return to_end_date(end_and_units)
+        return now()
+
+    def _get_snapshot(self, snapshot_name: str) -> t.Optional[Snapshot]:
+        return self.context_diff.snapshots.get(snapshot_name)
+
+    def _get_downstream_snapshots(self, snapshot: t.Union[str, Snapshot]) -> t.List[Snapshot]:
+        # We know that all downstream snapshots must exist since they can't be based on an external model since those
+        # only exist at the root level. Therefore we don't have to check for None here
+        return [self._get_snapshot(x) for x in self._dag.downstream(snapshot if isinstance(snapshot, str) else snapshot.name)]  # type: ignore
+
+    def _get_upstream_snapshots(self, snapshot: t.Union[str, Snapshot]) -> t.List[Snapshot]:
+        return [
+            snapshot
+            for snapshot in [
+                self._get_snapshot(snapshot_name)
+                for snapshot_name in self._dag.upstream(
+                    snapshot if isinstance(snapshot, str) else snapshot.name
```

Comment:
This expression comes up a lot. Should there be a function? `_snapshot_name` or something